File: backend/rag.py
import asyncio
from openai import AsyncOpenAI
from config import settings
from database import get_supabase

CHUNK_WORDS = 350
CHUNK_OVERLAP = 50

_openai: AsyncOpenAI | None = None


def _get_openai() -> AsyncOpenAI:
    global _openai
    if _openai is None:
        _openai = AsyncOpenAI(api_key=settings.OPENAI_API_KEY)
    return _openai


# ── Chunker ───────────────────────────────────────────────────────────────────

def chunk_text(text: str) -> list[str]:
    """Paragraph-aware sliding-window chunker (mirrors src/lib/rag.ts)."""
    paragraphs = [p.strip() for p in text.split("\n\n") if p.strip()]
    chunks: list[str] = []
    buffer: list[str] = []

    for para in paragraphs:
        words = para.split()
        buffer.extend(words)

        while len(buffer) >= CHUNK_WORDS:
            chunks.append(" ".join(buffer[:CHUNK_WORDS]))
            buffer = buffer[CHUNK_WORDS - CHUNK_OVERLAP:]

    if len(buffer) > 20:
        chunks.append(" ".join(buffer))

    return chunks


# ── Embeddings ────────────────────────────────────────────────────────────────

async def embed_text(text: str) -> list[float]:
    client = _get_openai()
    res = await client.embeddings.create(
        model="text-embedding-3-small",
        input=text[:8000],
    )
    return res.data[0].embedding


# ── Ingest ────────────────────────────────────────────────────────────────────

async def ingest_knowledge_item(
    org_id: str,
    title: str,
    content: str,
    item_type: str,
    metadata: dict | None = None,
) -> dict:
    supabase = get_supabase()

    # 1. Insert the raw item
    result = supabase.table("knowledge_items").insert({
        "organization_id": org_id,
        "title": title,
        "content": content,
        "item_type": item_type,
        "metadata": metadata or {},
    }).execute()

    item = result.data[0]

    # 2. Chunk + embed in batches of 10
    chunks = chunk_text(content)
    BATCH = 10

    for i in range(0, len(chunks), BATCH):
        batch = chunks[i:i + BATCH]
        embeddings = await asyncio.gather(*[embed_text(c) for c in batch])

        rows = [
            {
                "knowledge_item_id": item["id"],
                "organization_id": org_id,
                "chunk_text": batch[j],
                "chunk_index": i + j,
                "embedding": str(embeddings[j]),   # pgvector expects string repr
            }
            for j, _ in enumerate(batch)
        ]

        supabase.table("knowledge_embeddings").insert(rows).execute()

    return item


# ── Retrieve ──────────────────────────────────────────────────────────────────

async def retrieve_context(query: str, org_id: str, limit: int = 4) -> str:
    supabase = get_supabase()

    try:
        query_embedding = await embed_text(query)
    except Exception:
        return ""

    result = supabase.rpc("match_knowledge", {
        "query_embedding": str(query_embedding),
        "org_id": org_id,
        "match_count": limit,
        "match_threshold": 0.25,
    }).execute()

    if not result.data:
        return ""

    return "\n\n---\n\n".join(
        f"[{d['item_type'].upper()}: {d['title']}]\n{d['chunk_text']}"
        for d in result.data
    )

File: src/lib/rag.ts
import OpenAI from 'openai'
import { createAdminClient } from '@/lib/supabase/admin'

const CHUNK_WORDS = 350
const CHUNK_OVERLAP = 50

// ── Chunker ──────────────────────────────────────────────────────────────────

export function chunkText(text: string): string[] {
  // Split on paragraph breaks first, then by word count
  const paragraphs = text.split(/\n{2,}/).map(p => p.trim()).filter(Boolean)
  const chunks: string[] = []
  let buffer: string[] = []

  for (const para of paragraphs) {
    const words = para.split(/\s+/)
    buffer.push(...words)

    while (buffer.length >= CHUNK_WORDS) {
      chunks.push(buffer.slice(0, CHUNK_WORDS).join(' '))
      buffer = buffer.slice(CHUNK_WORDS - CHUNK_OVERLAP)
    }
  }

  if (buffer.length > 20) chunks.push(buffer.join(' '))
  return chunks
}

// ── Embeddings ───────────────────────────────────────────────────────────────

export async function embedText(text: string): Promise<number[]> {
  const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY! })
  const res = await openai.embeddings.create({
    model: 'text-embedding-3-small',
    input: text.slice(0, 8000), // model max
  })
  return res.data[0].embedding
}

// ── Store knowledge item + embeddings ────────────────────────────────────────

export async function ingestKnowledgeItem(params: {
  organizationId: string
  title: string
  content: string
  itemType: string
  metadata?: Record<string, unknown>
}) {
  const supabase = createAdminClient()

  // 1. Insert the raw item
  const { data: item, error: itemError } = await supabase
    .from('knowledge_items')
    .insert({
      organization_id: params.organizationId,
      title: params.title,
      content: params.content,
      item_type: params.itemType,
      metadata: params.metadata ?? {},
    })
    .select()
    .single()

  if (itemError) throw new Error(itemError.message)

  // 2. Chunk + embed in parallel (batch of 10)
  const chunks = chunkText(params.content)
  const BATCH = 10
  for (let i = 0; i < chunks.length; i += BATCH) {
    const batch = chunks.slice(i, i + BATCH)
    const embeddings = await Promise.all(batch.map(embedText))

    const rows = embeddings.map((embedding, j) => ({
      knowledge_item_id: item.id,
      organization_id: params.organizationId,
      chunk_text: batch[j],
      chunk_index: i + j,
      embedding: JSON.stringify(embedding),
    }))

    await supabase.from('knowledge_embeddings').insert(rows)
  }

  return item
}

// ── Retrieve relevant context for a query ────────────────────────────────────

export async function retrieveContext(
  query: string,
  orgId: string,
  limit = 4
): Promise<string> {
  const supabase = createAdminClient()

  let queryEmbedding: number[]
  try {
    queryEmbedding = await embedText(query)
  } catch {
    return ''
  }

  const { data, error } = await supabase.rpc('match_knowledge', {
    query_embedding: JSON.stringify(queryEmbedding),
    org_id: orgId,
    match_count: limit,
    match_threshold: 0.25,
  })

  if (error || !data || data.length === 0) return ''

  return data
    .map((d: { chunk_text: string; title: string; item_type: string }) =>
      `[${d.item_type.toUpperCase()}: ${d.title}]\n${d.chunk_text}`
    )
    .join('\n\n---\n\n')
}
